src/goto-programs/abstract-interpretation/gcse.cpp

```diff
@@ -0,0 +1,468 @@
+#include "goto-programs/goto_program.h"
+#include "irep2/irep2_expr.h"
+#include "irep2/irep2_type.h"
+#include "pointer-analysis/value_set_analysis.h"
+#include "util/std_code.h"
+#include "util/std_expr.h"
+#include <goto-programs/abstract-interpretation/gcse.h>
+#include <ostream>
+#include <sstream>
+#include <util/prefix.h>
+#include <fmt/format.h>
+// TODO: Do an points-to abstract interpreter
+std::shared_ptr<value_set_analysist> cse_domaint::vsa = nullptr;
+
+void cse_domaint::transform(
+  goto_programt::const_targett from,
+  goto_programt::const_targett to,
+  ai_baset &,
+  const namespacet &)
+{
+  const goto_programt::instructiont &instruction = *from;
+  switch(instruction.type)
+  {
+  case ASSIGN:
+  {
+    const code_assign2t &code = to_code_assign2t(instruction.code);
+    make_expression_available(code.source);
+    // Expressions that contain the target will need to be recomputed
+    havoc_expr(code.target, to);
+    // Target may be an expression as well
+    // TODO: skip recursive definitions only. For example, x = x + 1 should be skipped does not
+    // makes 'x' available. However, y = x + 1 does.
```

Comment:
I can't parse this TODO sentence, could you clarify?

Also, IIRC, having an `x = x` assignment will first make `x` available (if it's not already), and then it will remove `x` again from the available expressions, with the result that `x` no longer is available though no state has changed. Is that correct and intentional?

Reply:
Broken english, I will update the comment only with the exception.

Id expressions such as `x=x` should be fine indeed. The issue of `x = x + 1` is that all previous dependencies of `x` are no longer available. If I add it back though, it will cause issues during the merge operation.